app/v1/support/DesResponseMappingSupport.scala

```diff
@@ -33,6 +35,28 @@ trait DesResponseMappingSupport {
     }
   }
 
+  final def validateCodingOutResponse[T](desResponseWrapper: ResponseWrapper[T], taxYear: String)
+                                        (implicit currentDate: CurrentDate): Either[ErrorWrapper, ResponseWrapper[T]] = {
+    desResponseWrapper.responseData match {
+      case retrieveCodingOutDetailsResponse: RetrieveCodingOutResponse if TaxYearNotEndedValidation.validate(taxYear).isEmpty
+        && checkIfIdDoesntExists(retrieveCodingOutDetailsResponse) =>
+        Left(ErrorWrapper(desResponseWrapper.correlationId, DownstreamError))
+      case _ => Right(desResponseWrapper)
+    }
+  }
+
+  def checkIfIdDoesntExists(response: RetrieveCodingOutResponse): Boolean = {
+
+    response.taxCodeComponents.flatMap(_.debt.map(_.map(_.id.isEmpty))).head.head ||
+    response.taxCodeComponents.flatMap(_.inYearAdjustment.map(_.id.isEmpty)).head ||
+    response.taxCodeComponents.flatMap(_.payeUnderpayment.map(_.map(_.id.isEmpty))).head.head||
+    response.taxCodeComponents.flatMap(_.selfAssessmentUnderpayment.map(_.map(_.id.isEmpty))).head.head||
+    response.unmatchedCustomerSubmissions.flatMap(_.debt.map(_.map(_.id.isEmpty))).head.head ||
+    response.unmatchedCustomerSubmissions.flatMap(_.inYearAdjustment.map(_.id.isEmpty)).head ||
+    response.unmatchedCustomerSubmissions.flatMap(_.payeUnderpayment.map(_.map(_.id.isEmpty))).head.head ||
+    response.unmatchedCustomerSubmissions.flatMap(_.selfAssessmentUnderpayment.map(_.map(_.id.isEmpty))).head.head
+  }
```

Comment:
```suggestion
  def idsMissing(response: RetrieveCodingOutResponse): Boolean = {
    import response._

    lazy val taxCodeComponentsIdsMissing = taxCodeComponents.fold(false)(taxCodeComponents => {
      import taxCodeComponents._

      debt.fold(false)(debt => debt.exists(_.id.isEmpty)) ||
        inYearAdjustment.fold(false)(inYearAdjustment => inYearAdjustment.id.isEmpty) ||
        payeUnderpayment.fold(false)(payeUnderpayment => payeUnderpayment.exists(_.id.isEmpty)) ||
      selfAssessmentUnderpayment.fold(false)(selfAssessmentUnderpayment => selfAssessmentUnderpayment.exists(_.id.isEmpty))
    })

    lazy val unmatchedCustomerSubmissionsIdsMissing = unmatchedCustomerSubmissions.fold(false)(unmatchedCustomerSubmissions => {
      import unmatchedCustomerSubmissions._

      debt.fold(false)(debt => debt.exists(_.id.isEmpty)) ||
        inYearAdjustment.fold(false)(inYearAdjustment => inYearAdjustment.id.isEmpty) ||
        payeUnderpayment.fold(false)(payeUnderpayment => payeUnderpayment.exists(_.id.isEmpty)) ||
        selfAssessmentUnderpayment.fold(false)(selfAssessmentUnderpayment => selfAssessmentUnderpayment.exists(_.id.isEmpty))
    })

    taxCodeComponentsIdsMissing && unmatchedCustomerSubmissionsIdsMissing
  }
```
The code here is a little hard to follow. Could you try to clean it up a little? I've included a possible suggestion but i'm sure there are better ways of doing it.